Extend the readme so we can have a preview of the documentation (#66)
metallb/frr-k8s: README.md

# frr-k8s

A kubernetes based daemonset that exposes a subset of the FRR API in a kubernetes compliant manner.

The rationale behind the creation of this project is to allow multiple actors to share a single FRR instance
running on kubernetes nodes. More details can be found in the [metallb design proposal](https://github.com/metallb/metallb/blob/main/design/splitfrr-proposal.md).

## Description

> The frr-k8s is deployed as a daemonset on Kubernetes clusters' nodes.

Author: I'll leave it as just "Frr-k8s"

Each FRR-K8s pod embeds an instance of [FRR](https://frrouting.org/).

In order to configure each FRR, the various instances of the `FRRConfiguration` CRD are merged together and converted in
a single configuration for the FRR container.

<img src="docs/architecture.png" alt="architecture" width="600"/>

## Installing

The frr-k8s daemon provides 3 deployment methods.

### All in one manifests

> This is the simples way to deploy and test frr-k8s, although the less configurable one.
> 
> Just

Reviewer: ```suggestion
This is the simplest way to deploy and test frr-k8s, although the least configurable one:
```

```bash
helm install frr-k8s frr-k8s/frr-k8s -f values.yaml
```

**NOTE:** The frr-k8s pod requires elevated permission in order to perform its network functionalities.

If you are using frr-k8s with a kubernetes version that enforces pod security admission (which is beta in k8s 1.23), the namespace frr-k8s is deployed to must be labelled with:

```yaml
  labels:
    pod-security.kubernetes.io/enforce: privileged
    pod-security.kubernetes.io/audit: privileged
    pod-security.kubernetes.io/warn: privileged
```

## Configuring FRR-K8s

The CRD to configure `FRR-K8s` is the `FRRConfiguration`. A comprehensive documentation of the API can be found in
the [API Docs file](./API-DOCS.md)

### BGP Configuration

The `BGP` section of the FRRConfiguration spec is in charge of configuring the details related to the BGP protocol.

#### Configuring one (or multiple) routers

It is possible to configure multiple routers, one for each VRF.

For each router, you can define the ASN and the VRF it belongs to, plus a list of prefixes to announce
and a list of BGP neighbors to connect to:

```yaml
spec:
  bgp:
    routers:
    - asn: 64512

>       neighbors:

Reviewer: maybe also add a `toAdvertise` since we declare prefixes?

Author: toAdvertise is expanded later

      - address: 172.30.0.3
        asn: 4200000000
        ebgpMultiHop: true
        port: 180
      - address: 172.18.0.6
        asn: 4200000000
        port: 179
      prefixes:
      - 192.168.2.0/24
      - 192.169.2.0/24
```

#### Associating a neighbor with BFD

It is possible to define various BFD profiles (in the `bgp` section of the spec) and associate them to a neighbor.
By doing so, the BGP session is backed up by BFD and thus FRR is able to detect link failures quickly.

```yaml
spec:
  bgp:
    routers:
    - asn: 64512
      neighbors:
      - address: 172.30.0.3
        asn: 64512
        port: 180
        bfdProfile: defaultprofile
    bfdProfiles:
      - name: defaultprofile
```

> #### Advertising a prefix to a given neighbor

Reviewer: or put this before the bfd part / I think this should just be a part of the first example without split yamls  and beneath it the "By default ..." part (and currently the split yamls look malformed

By default, the prefixes configured as part of a router configuration are not advertised. In order to advertise them,
the `toAdvertise` section must be filled.

A subset of the prefixes can be advertised:

```yaml
    - address: 172.18.0.6
    asn: 4200000000
    port: 179
    toAdvertise:
        allowed:
        prefixes:
        - 192.168.2.0/24
```

or all the prefixes configured in the router can be advertised:

```yaml
    - address: 172.30.0.3
    asn: 4200000000
    ebgpMultiHop: true
    port: 180
    toAdvertise:
        allowed:
        mode: all
```

#### Receiving prefixes from a given neighbor

By default, no prefixes advertised by a neighbour are processed.

In order to process them, a `toReceive` section must be filled.

A subset of the prefixes can be received:

> ```yaml

Reviewer: the split yamls look malformed + I think it is more readable as a whole config

        - 192.169.2.0/24
```

or FRR can be configured to handle all the prefixes announced:

```yaml
    - address: 172.18.0.5
    asn: 64512
    port: 179
    toReceive:
        allowed:
        mode: all
```

### Adding a raw configuration

In order to facilitate experimentation and to fill gaps quickly, it is possible to set a piece of raw
FRR configuration that is appended to the rendered one. In order to do so, the `rawConfig` field
must be set:

```yaml
apiVersion: frrk8s.metallb.io/v1beta1
kind: FRRConfiguration
metadata:
  name: test1
  namespace: frr-k8s-system
spec:
  raw:
    priority: 5
    rawConfig: |-
      router bgp 64512
        neighbor 172.18.0.5 remote-as 4200000000
        neighbor 172.18.0.5 timers 0 0
        neighbor 172.18.0.5 port 179
```

The priority field sets the order in case multiple configurations are merged together.

### Node selector

A node selector field drives the nodes where the given configuration is applied.

```yaml
apiVersion: frrk8s.metallb.io/v1beta1
kind: FRRConfiguration
metadata:
  name: test
  namespace: frr-k8s-system
spec:
  bgp:
    routers:
    - asn: 64512
  nodeSelector:
    labelSelector:
      foo: "bar"
```

> No node selector means the configuration is applied to all the nodes where the daemon is running:

Reviewer: ```suggestion
By not setting the node selector the configuration is applied to all nodes where the daemon is running.
```

### How multiple configurations are merged together

Multiple actors may add configurations selecting the same node. In this case, the configurations are merged together.

The way `frr-k8s` configurability is intended is, each configuration can only extend others. This means that it is
allowed (and possible) to add a new neighbor to a router, or to advertise an additional prefix to a neighbor, but there
is no way to remove a component added by another configuration.

> 

Reviewer: I'd add something about each configuration being independent and cannot rely on others, and in case of overlap the more permissive one takes precedence (such as allowall in incoming routes)

#### Configuration conflicts

There are configurations that may conflict between themselves, leading to errors.

> This include for example:

Reviewer: nit: includes

- different ASN for the same router (in the same VRF)
- different ASN for the same neighbor (with the same ip / port)
- multiple BFD profiles with the same name but different values

When the daemon finds an invalid configuration state of a given node, it will report the configuration as invalid and it will
leave the previous valid FRR configuration.

> #### Merging

Reviewer: or put this higher :smile:

If the configurations to be applied to a given node are compatible, merging works by:

- combining all the routers
- for each router, merging all the prefixes and the neighbors
- for each neighbor, merging all the filters*

\*: a more permissive filter has precedence over a less permissive one, so accepting some prefixes has precedence over
not accepting any, and accepting all the prefixes has precedence over accepting some.

By doing so, it is possible to:

- Extend the set of ips we want to advertise to a given neighbor
- Adding an extra neighbor with its set of IPs
- Extend the set of ips we want to associate a given community to
- Allow incoming routes for a given neighbor

**Each configuration must be self contained!** This means for example that it's not possible to allow prefixes that are not defined in the router section, leveraging prefixes coming from another configuration.

## Checking the status of each FRR instance

> The `frrNodeState` crd expose the status of the FRR instance running on each node.

Reviewer: ```suggestion
The `FRRNodeState` resource exposes the status of the FRR instance running on each node.
```

This includes:

> - the desired FRR config, as result of the translation from the result of all the `FRRConfiguration`s for the given node
> - the current FRR config, which is the configuration the FRR instance is currently running with
> - the status of the last configuration update operation
> - the status of the last translation between the `FRRConfiguration`s and the frr configuration  

Reviewer: ```suggestion
- desiredConfig: the desired FRR config, as the result of translating all of the `FRRConfiguration`s for the given node.
- runningConfig: the current FRR running config, which is the configuration the FRR instance is currently running with
- lastReloadResult: the status of the last configuration update operation by FRR, contains "success" or an error.
- lastConversionResult: the status of the last translation between the `FRRConfiguration`s resources and FRR's configuration, contains "success" or an error.
```

## MetalLB Integration

> The primary reason of existance of this project is to allow users to leverage FRR while MetalLB still configures it.

Reviewer: ```suggestion
This project was created to allow users to configure FRR while MetalLB does.
```

# frr-k8s

A kubernetes based daemonset that exposes a subset of the FRR API in a kubernetes compliant manner.

The rationale behind the creation of this project is to allow multiple actors to share a single FRR instance
running on kubernetes nodes. More details can be found in the [metallb design proposal](https://github.com/metallb/metallb/blob/main/design/splitfrr-proposal.md).

## Description

> Frr-k8s is deployed as a daemonset on Kubernetes clusters' nodes.

Reviewer: nit: `Frr-x` -> `FRR-x`

Each FRR-K8s pod embeds an instance of [FRR](https://frrouting.org/).

Tthe `FRRConfiguration` CRD is the way a cluster administrator can configure each `frr-k8s` instance running on the various nodes.

Each instance will collect all the `FRRConfiguration`s applied to the node it's running on, will merge them and produce the
corresponding FRR configuration.

Additionally, each `frr-k8s` instance will produce a `FRRNodeState` instance exposing the status of the FRR
instance running on each node.

<img src="docs/architecture.png" alt="architecture" width="600"/>

## Installing

The frr-k8s daemon provides 3 deployment methods.

### All in one manifests

> This is the simples way to deploy and test frr-k8s, although the less configurable one.

Reviewer: `simples` -> `simplest`, `less` -> `least`

    - asn: 64512
      neighbors:
      - address: 172.30.0.3
        asn: 4200000000
        ebgpMultiHop: true
        port: 180
      - address: 172.18.0.6
        asn: 4200000000
        port: 179
```

#### Advertising a prefix to a given neighbor

By default, the prefixes configured as part of a router configuration are not advertised. In order to advertise them,
the `toAdvertise` section must be filled.

A subset of the prefixes can be advertised:

```yaml
spec:
  bgp:
    routers:
    - asn: 64512
      neighbors:
      - address: 172.30.0.3
        asn: 4200000000
        ebgpMultiHop: true
        port: 180
        toAdvertise:
          allowed:
            prefixes:
            - 192.168.2.0/24
      prefixes:
        - 192.168.2.0/24
        - 192.169.2.0/24
```

or all the prefixes configured in the router can be advertised:

```yaml
spec:
  bgp:
    routers:
    - asn: 64512
      neighbors:
      - address: 172.30.0.3
        asn: 4200000000
        ebgpMultiHop: true
        port: 180
        toAdvertise:
          allowed:
            mode: all
      prefixes:
        - 192.168.2.0/24
        - 192.169.2.0/24
```

#### Receiving prefixes from a given neighbor

> By default, no prefixes advertised by a neighbour are processed.

Reviewer: nit: `neighbour` -> `neighbor`

In order to process them, a `toReceive` section must be filled.

A subset of the prefixes can be received:

> ```yaml

Reviewer: this should either be `spec: ...` like the others / a complete yaml

    - address: 172.18.0.5
    asn: 64512
    port: 179
    toReceive:
      allowed:
        prefixes:
        - prefix: 192.168.2.0/24

>           le: 32

Reviewer: I'd keep one without le/ge, explain/reference frr docs what le/ge do

        - prefix: 192.169.2.0/24
          ge: 25
          le: 28
```

or FRR can be configured to handle all the prefixes announced:

> ```yaml

Reviewer: same

leave the previous valid FRR configuration.

#### Merging

If the configurations to be applied to a given node are compatible, merging works by:

- combining all the routers
- for each router, merging all the prefixes and the neighbors
- for each neighbor, merging all the filters*

\*: a more permissive filter has precedence over a less permissive one, so accepting some prefixes has precedence over
not accepting any, and accepting all the prefixes has precedence over accepting some.

By doing so, it is possible to:

- Extend the set of ips we want to advertise to a given neighbor
- Adding an extra neighbor with its set of IPs
- Extend the set of ips we want to associate a given community to
- Allow incoming routes for a given neighbor

**Each configuration must be self contained!** This means for example that it's not possible to allow prefixes that are not defined in the router section, leveraging prefixes coming from another configuration.

## Checking the status of each FRR instance

The `FRRNodeState` resource exposes the status of the FRR instance running on each node.

> This includes:

Reviewer: I think we should add these to the docs comment commit as well

> - desiredConfig: the desired FRR config, as the result of translating all of the `FRRConfiguration`s for the given node.

Reviewer: nit: maybe wrap the field names in ``

> - runningConfig: the current FRR running config, which is the configuration the FRR instance is currently running with

Reviewer: nit: missing `.`

- lastReloadResult: the status of the last configuration update operation by FRR, contains "success" or an error.
- lastConversionResult: the status of the last translation between the `FRRConfiguration`s resources and FRR's configuration, contains "success" or an error.

## MetalLB Integration

> This project was created as a solution to allow users to leverage the same FRR instance used by MetalLB does.

Reviewer: remove `does`
